### PR TITLE
[WIP adding tests] Prompt rect for autocorrection (and multistage text input)  

### LINKED ISSUE #12920: Autocorrect tooltips don't appear on iOS
## Steps to Reproduce

When using a native iOS app without text prediction enabled, there is a pop-up that indicates when autocorrections are going to occur, and when backspacing on a correction a pop-up appears that gives alternative choices for the word.

In flutter, the autocorrections happen silently and require backspacing into the word to make the correction.  See the gif below for a comparison of the behavior.

Flutter should either integrate with the iOS autocorrection system to provide the popups or emulate them.  This is a pretty big usability concern which makes it more difficult to enter text because the corrections are harder to notice and harder to fix.

text_correction

## Logs

N/A, this is a design issue.

## Flutter Doctor

```
$ flutter doctor
Building flutter tool...
[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-US, channel master)
    • Flutter at /Users/jwm/Projects/flutter
    • Framework revision 337150308c (2 days ago), 2017-11-05 19:39:49 -0800
    • Engine revision e059cc0258
    • Tools Dart version 1.25.0-dev.11.0

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set $ANDROID_HOME to that location.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.0.1, Build version 9A1004
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✗] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2017.2.5)
    • Flutter plugin version 18.2
    • Dart plugin version 172.4343.25

[✓] Connected devices
    • iPhone 7 Plus • DF665BBF-F7D6-4B75-8FCA-E092AD4AD6DC • ios • iOS 11.0 (simulator)
```

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description
Engine counterpart: https://github.com/flutter/engine/pull/9539

Display prompt rectangles when `UIKit` decides it's appropriate to do so (because of autocorrection/multistage text input, etc.).

Use `-[UITextInput firstRectForRange:]` as an indicator of when and where the prompt rectangle should show up, and dismiss the prompt rectangle eagerly in `renderEditable.performLayout`, since there doesn't seem to be a good way to tell when the rectangle should be dismissed.

## Known Fidelity Issues

- If one types "f" after "asd" , the autocorrection prompt rectangle for "asd" will go away then reappear for "asdf", instead of just extending the old rectangle to cover the new word.

A gif showcasing the difference between autocorrection & multistage text input:
ezgif-4-6e01122435eb

## Related Issues

Fixes #12920

## Tests

To Be Added.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by justinmc — APPROVED** [review]
LGTM, just minor stuff below.  I tried running this and didn't see the rectangle. Is there anything to do to enable it?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1523,6 +1562,13 @@ class RenderEditable extends RenderBox {
     _layoutText(constraints.maxWidth);
     _caretPrototype = _getCaretPrototype;
     _selectionRects = null;
+
+    // Only dismiss the rect if not composing multistage text, in order to prevent
+    // flickering. Unfortunately this hack leaks iOS implementation detail.
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart]: This doesn't seem terrible to me...

File: packages/flutter/lib/src/services/text_input.dart [hunk 2]
```diff
@@ -347,6 +347,20 @@ enum TextCapitalization {
   none,
 }
 
+/// Indicates what caused the text input prompt rectangle to appear.
+enum TextInputPromptRectAppearCause {
+  /// The prompt rectangle appeared because of a unknown reason.
+  unspecified,
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_input.dart]: Makes sense. Removed `unspecified`. If we want to add a new cause to the enum we break the API anyways.

File: packages/flutter/lib/src/services/text_input.dart [hunk 3]
```diff
@@ -347,6 +347,17 @@ enum TextCapitalization {
   none,
 }
 
+/// Indicates what caused the text input prompt rectangle to appear.
+enum TextInputPromptRectAppearCause {
+  /// The prompt rectangle appeared to indicate the range of text that will be
+  /// affected by autocorrection.
+  autocorrection,
```
Comment (Hixie) [inline packages/flutter/lib/src/services/text_input.dart]: might be worth adding some examples to these docs (not like code examples, i mean like user interaction examples), to clarify what they're about

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -1000,6 +1004,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     }
   }
 
+  @override
+  void showPromptRect(int start, int end, TextInputPromptRectAppearCause cause) {
+    if (cause == TextInputPromptRectAppearCause.autocorrection && !widget.autocorrect) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/editable_text.dart]: per style guide, prefer switch statements when comparing enums

File: dev/integration_tests/ui/test_driver/driver_test.dart [hunk 5]
```diff
@@ -3,8 +3,10 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:io' show Platform;
```
Comment (goderbauer) [inline dev/integration_tests/ui/test_driver/driver_test.dart]: Cirrus is complaining that this import and the other new one below is unused. Remove?

**Pull request closed without merging**